interfaces/apparmor/backend.go

```diff
@@ -426,6 +436,59 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	return errUnload
 }
 
+func (b *Backend) SetupMany(snaps []*snap.Info, confinement func(snapName string) interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) []error {
+	var allChangedPaths, allUnchangedPaths, allRemovedPaths []string
+	var fallback bool
+	for _, snapInfo := range snaps {
+		opts := confinement(snapInfo.InstanceName())
+		changedPaths, removedPaths, unchangedPaths, err := b.prepareProfiles(snapInfo, opts, repo)
+		if err != nil {
+			fallback = true
+			break
+		}
+		allChangedPaths = append(allChangedPaths, changedPaths...)
+		allUnchangedPaths = append(allUnchangedPaths, unchangedPaths...)
+		allRemovedPaths = append(allRemovedPaths, removedPaths...)
+	}
+
+	if !fallback {
+		var errReloadChanged error
+		timings.Run(tm, "load-profiles-many[changed]", fmt.Sprintf("load changed security profiles of %d snaps", len(snaps)), func(nesttm timings.Measurer) {
+			errReloadChanged = loadProfiles(allChangedPaths, dirs.AppArmorCacheDir, skipReadCache|conserveCPU)
+		})
+
+		var errReloadOther error
+		timings.Run(tm, "load-profiles-many[unchanged]", fmt.Sprintf("load unchanged security profiles %d snaps", len(snaps)), func(nesttm timings.Measurer) {
+			errReloadOther = loadProfiles(allUnchangedPaths, dirs.AppArmorCacheDir, conserveCPU)
+		})
+		errUnload := unloadProfiles(allRemovedPaths, dirs.AppArmorCacheDir)
+		if errReloadChanged != nil {
+			logger.Noticef("failed to reload changed profiles: %s", errReloadChanged)
+			fallback = true
+		}
+		if errReloadOther != nil {
+			logger.Noticef("failed to reload unchanged profiles: %s", errReloadOther)
+			fallback = true
+		}
+		if errUnload != nil {
+			logger.Noticef("failed to unload profiles: %s", errUnload)
+			fallback = true
+		}
+	}
+
+	var errors []error
+	// if an error was encountered when processing all profiles at once, re-try them one by one
+	if fallback {
```

Comment:
I'm definitely interested in @jdstrand input whether there might be problems with this approach

Reply:
> I'm definitely interested in @jdstrand input whether there might be problems with this approach

Actually, I really like the approach and agree with @stolowski that we shouldn't ever hit this case except in the face of snapd bugs or FS corruption. The policy should always compile and if it doesn't, that is an important bug to fix (this is why we have the interfaces-many* tests). Compiling in parallel is in general a good thing and this is a nice way to load everything and report back on issues. Plus, anything that did compile will already have the updated profiles on disk, with the regenerated cache and so the extra apparmor_parser calls are going to be essentially as fast as we can fork/exec, do two stat()s and reading the cache file off the disk for each profile. I also verified back to xenial that apparmor_parser will load everything it can and not bail on the first error. Considering how seldom we should ever fallback, this seems reasonable.

That said, I'd like to see a test case that corrupts an apparmor profile and demonstrates that the fallback code works as expected and shows the error for the bad profile. This will be tricky because corrupting the profile after snapd writes it be before apparmor_parser parses it will mean that the fallback will rewrite it to its pre-corrupted state. Creative parser mocking instead is possible.

Reply:
Thank you! I've added unit tests covering the fallback scenario, hower a real life spread test is indeed tricky; I'll see if I can come up with something.

Reply:
Spread test available now.